src/NuGet.Protocol.Core.v2/DownloadResourceV2.cs

```diff
@@ -66,5 +69,211 @@ public DownloadResourceV2(V2Resource resource)
+            {
+                throw new NuGetProtocolException(Strings.FormatProtocol_FailedToDownloadPackage(
+                    package,
+                    dataServiceRepo.Source),
+                    ex);
+            }
+
+            return null;
+        }
+
+        private static DownloadResourceResult DownloadFromIdentity(
+            PackageIdentity identity, 
+            IPackageRepository repository, 
+            CancellationToken token)
+        {
+            var version = SemanticVersion.Parse(identity.Version.ToString());
+            var dataServiceRepo = repository as DataServicePackageRepository;
+
+            if (dataServiceRepo != null)
+            {
+                // Clone the repo to allow for concurrent calls
+                var sourceUri = new Uri(dataServiceRepo.Source);
+                dataServiceRepo = new DataServicePackageRepository(sourceUri);
+
+                var package = dataServiceRepo.FindPackage(identity.Id, version);
+                token.ThrowIfCancellationRequested();
+
+                // For online sources get the url and retrieve it with cancel support
+                var dataServicePackage = package as DataServicePackage;
+                var url = dataServicePackage.DownloadUrl;
+
+                var downloadedPackage = DownloadToMachineCache(
+                    MachineCache.Default,
+                    identity,
+                    dataServiceRepo,
+                    url,
+                    token);
+
+                if (downloadedPackage != null)
+                {
+                    return new DownloadResourceResult(downloadedPackage.GetStream());
+                }
+            }
+            else
+            {
+                var package = repository.FindPackage(identity.Id, version);
+
+                // Use a folder reader for unzipped repos
+                if (repository is UnzippedPackageRepository)
+                {
+                    var packagePath = Path.Combine(repository.Source, identity.Id + "." + version);
+                    var directoryInfo = new DirectoryInfo(packagePath);
+                    if (directoryInfo.Exists)
+                    {
+                        return new DownloadResourceResult(
+                            package.GetStream(),
+                            new PackageFolderReader(directoryInfo));
+                    }
+                }
+
+                return new DownloadResourceResult(package.GetStream());
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// True if the given package matches hash
+        /// </summary>
+        private static bool MatchPackageHash(IPackage package, string hash)
+        {
+            var hashProvider = new CryptoHashProvider("SHA512");
+
+            return package != null && package.GetHash(hashProvider).Equals(hash, StringComparison.OrdinalIgnoreCase);
+        }
+
+        private static IPackage DownloadToMachineCache(
```

Comment:
The v2 machine cache has never had mutexes like the global folder now does. It downloads to a tmp file and swaps the file after the download completes. It would be nice to have them if we wish to devote more cycles to NuGet.Core.